include/deal.II/lac/block_indices.h

```diff
@@ -333,11 +333,11 @@ BlockIndices::global_to_local(const size_type i) const
   Assert(i < total_size(), ExcIndexRangeType<size_type>(i, 0, total_size()));
   Assert(n_blocks > 0, ExcLowerRangeType<size_type>(i, size_type(1)));
 
-  unsigned int block = n_blocks - 1;
-  while (i < start_indices[block])
-    --block;
+  // start_indices[0] == 0 so we might as well start from the next one
+  const auto it =
+    --std::upper_bound(++start_indices.begin(), start_indices.end(), i);
```

Comment:
what happens if there is only one block? then `begin()==end()` and the function must return `end()`. Is it allowed to take `--end()`?

Reply:
if there is just one block then start_indices is of size 2 and begin is still not equal to end.

Reply:
So decrementing should be ok.

Reply:
Ah, so we store a dummy element at the end?

Reply:
Global starting index of each vector. The last and redundant value is the total number of entries.